include/cpp_core/serial.h

```diff
@@ -0,0 +1,393 @@
+     *                     first.
+     * @return             Bytes written (including the newline) or a negative
+     *                     error code.
+     */
+    MODULE_API auto serialWriteLine(
+        int64_t     handle,
+        const void *buffer,
+        int         buffer_size,
+        int         timeout_ms,
+        int         multiplier
+    ) -> int;
+
+    /**
+     * @brief Read until a specific byte sequence appears.
+     *
+     * Works like serialReadUntil() but supports an arbitrary terminator string.
+     * The terminator is included in the returned data.
+     *
+     * @param handle       Port handle.
+     * @param buffer       Destination buffer.
+     * @param buffer_size  Capacity of @p buffer in bytes.
+     * @param timeout_ms   Base timeout per byte in milliseconds (first byte
+     *                     uses this value; each additional byte uses
+     *                     `timeout_ms * multiplier`).
+     * @param multiplier   Factor applied to the timeout for subsequent bytes.
+     * @param sequence     Pointer to the terminating byte sequence (must not be `nullptr`).
+     * @return             Bytes read (including the terminator) or a negative error code.
+     */
+    MODULE_API auto serialReadUntilSequence(
+        int64_t handle,
+        void   *buffer,
+        int     buffer_size,
+        int     timeout_ms,
+        int     multiplier,
+        void   *sequence
+    ) -> int;
+
+    /**
+     * @brief Total number of bytes transmitted since the port was opened.
+     *
+     * @param handle Port handle.
+     * @return       Total number of bytes written.
+     */
+    MODULE_API auto serialOutBytesTotal(int64_t handle) -> int64_t;
+
+    /**
+     * @brief Total number of bytes received since the port was opened.
+     *
+     * @param handle Port handle.
+     * @return       Total number of bytes read.
+     */
+    MODULE_API auto serialInBytesTotal(int64_t handle) -> int64_t;
+
+    /**
+     * @brief Wait until the operating-system driver has physically sent every queued byte.
+     *
+     * The function blocks the *calling thread* until the device driver reports
+     * that the transmit FIFO is **empty** – i.e. all bytes handed to previous
+     * `serialWrite*` calls have been shifted out on the wire.  It does *not*
+     * flush higher-level protocol buffers you may have implemented yourself.
+     *
+     * Typical use-case: ensure a complete command frame has left the UART
+     * before toggling RTS/DTR or powering down the device.
+     *
+     * @code{.c}
+     * // Send a frame and make sure it actually hits the line
+     * serialWrite(h, frame, frame_len, 50, 1);
+     * if (serialDrain(h) < 0) {
+     *     fprintf(stderr, "drain failed\n");
+     * }
+     * @endcode
+     *
+     * @param handle Port handle.
+     * @return       0 on success; negative ::cpp_core::StatusCodes value on error
+     *               (e.g. kInvalidHandleError).
```

Comment:
If you provide an error for this return, please also provide one for every function that possibly returns an error.

```suggestion
     * @return       0 on success; negative ::cpp_core::StatusCodes value on error.
```